### PR TITLE
Added a git guide for new members

### PR DESCRIPTION
**Pull request by KipHamiltons**
https://deploy-preview-54--nubook.netlify.app/guides/general/git

### CODE REVIEW

File: src/book/04-general/01-git.mdx [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+
+## Prerequisites
+
+This guide assumes you know the basics of git and basic shell commands.
+
+## Getting Code Into a Repository
+
+Suppose you have a file you have made which you would like to add to a NUbots repository, called `my_file.cpp`, as part of your project to make the robot stand up.
+
+In terminal, change directory to where you have the repo cloned.
```
Comment (ysims) [inline src/book/04-general/01-git.mdx:17]:
```suggestion
In the terminal, change directory to where you have the repo cloned.
```

File: src/book/04-general/01-git.mdx [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+### Select Branch
+
+If you are working within an existing project, then you need to find out which branch the work is on. To list the branches, use `git branch`. Once you can see which branch you to push to swap to it:
```
Comment (ysims) [inline src/book/04-general/01-git.mdx:21]:
```suggestion
If you are working within an existing project, then you need to find out which branch the work is on. To list the branches, use `git branch`. When you have found the branch you want to push to, swap to it:
```

File: src/book/04-general/01-git.mdx [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+git checkout <branch name>
+```
+
+If you are not working on an existing project, you need to make a branch to commit your changes to. The format for branch names is `surname/project_name`. So for this example it could be `citizen/stand_up`.
+
+### Making a Pull Request
+
+Once you have a set of commits on your branch which you think should be added to the main codebase, you can make a Pull Request (PR). PRs are where users ask the repository to pull their branch into the `master` branch.
+
+Github has documentation about about PRs [here.](#https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/creating-a-pull-request)
```
Comment (ysims) [inline src/book/04-general/01-git.mdx:33]:
```suggestion
GitHub has documentation about PRs [here.](#https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/creating-a-pull-request)
```

File: src/book/04-general/01-git.mdx [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+
+Title your PR with a description of the changes that it will make. For this example we could use `Add stand-up script`. If you need to give more detail about what the code does, put that as a comment.
+
+To be approved, you should add reviewers to the pull request. If you have a project mentor, they should be a reviewer, as well as someone else who knows about that part of the codebase. Once these people approve the changes, they can be merged in.
```
Comment (ysims) [inline src/book/04-general/01-git.mdx:37]:
```suggestion
To be approved, you should add reviewers to the pull request. If you have a project mentor, they should be a reviewer, as well as someone else who knows about that part of the codebase. Once the reviewers have approved the changes, the changes can be merged into the master branch.
```

File: src/book/04-general/01-git.mdx [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+
+### Code Review
+
+The reviewers will give constructive feedback about your code, by making comments on code snippets from your changes. You can respond to these on github. They can also directly suggest changes to specific sections. Additional commits you make to the branch will update the PR and will reset their review of changed sections.
```
Comment (ysims) [inline src/book/04-general/01-git.mdx:41]:
```suggestion
The reviewers will give constructive feedback about your code, by making comments on code snippets from your changes. You can respond to these on GitHub. They can also directly suggest changes to specific sections. Additional commits you make to the branch will update the PR and will reset their review of changed sections.
```

File: src/book/04-general/01-git.mdx [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+section: Guides
+chapter: General
+title: Git
+description: Contributing to a NUbots project with git.
+slug: /guides/general/git
+---
+
+## Prerequisites
+
+This guide assumes you know the basics of git and basic shell commands.
```
Comment (ysims) [inline src/book/04-general/01-git.mdx:11]: Maybe specify what is basic.

File: src/book/04-general/01-git.mdx [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+
+## Getting Code Into a Repository
+
+Suppose you have a file you have made which you would like to add to a NUbots repository, called `my_file.cpp`, as part of your project to make the robot stand up.
```
Comment (ysims) [inline src/book/04-general/01-git.mdx:15]: The wording is a little hard to follow. Maybe something like
```suggestion
Let's suppose you have made a file, which you would like to add to a NUbots repository, called `my_file.cpp`. Suppose this file is part of your project to make the robot stand up.
```

File: src/book/04-general/01-git.mdx [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+
+In the terminal, change directory to where you have the repo cloned.
```
Comment (ysims) [inline src/book/04-general/01-git.mdx:17]:
```suggestion
In the terminal, change directory to the location where the repository is cloned.
```

File: src/book/04-general/01-git.mdx [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+
+### Select Branch
+
+If you are working within an existing project, then you need to find out which branch the work is on. To list the branches, use `git branch`. When you have found the branch you want to push to, swap to it:
```
Comment (ysims) [inline src/book/04-general/01-git.mdx:21]:
```suggestion
If you are working within an existing project, then you need to find out which branch the work is on. To list the branches, use the command `git branch`. When you have found the branch you want to push to, swap to it using the following command:
```

File: src/book/04-general/01-git.mdx [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+
+```sh
+git checkout <branch name>
+```
+
+If you are not working on an existing project, you need to make a branch to commit your changes to. The format for branch names is `surname/project_name`. So for this example it could be `citizen/stand_up`.
```
Comment (ysims) [inline src/book/04-general/01-git.mdx:27]:
```suggestion
Where `<branch name>` corresponds to the branch you found in the last step. 

If you are not working on an existing project, you need to make a branch to commit your changes to. The format for branch names is `surname/project_name`. So for this example it could be `citizen/stand_up`, if your surname is Citizen.
```

File: src/book/04-general/01-git.mdx [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+
+Github has documentation about PRs [here.](#https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/creating-a-pull-request)
+
+Title your PR with a description of the changes that it will make. For this example we could use `Add stand-up script`. If you need to give more detail about what the code does, put that as a comment.
+
+To be approved, you should add reviewers to the pull request. If you have a project mentor, they should be a reviewer, as well as someone else who knows about that part of the codebase. Once the reviewers have approved the changes, the changes can be merged into the master branch.
+
+### Code Review
+
+The reviewers will give constructive feedback about your code, by making comments on code snippets from your changes. You can respond to these on Github. They can also directly suggest changes to specific sections. Additional commits you make to the branch will update the PR and will reset their review of changed sections.
```
Comment (ysims) [inline src/book/04-general/01-git.mdx:41]:
```suggestion
The reviewers will give constructive feedback about your code, by making comments on code snippets from your changes. You can respond to these on GitHub. They can also directly suggest changes to specific sections. Additional commits you make to the branch will update the PR and will reset reviews of the changed sections.
```

File: src/book/04-general/01-git.mdx [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+
+Once the reviewers are happy with it, and all of the comments are resolved, you can `Squash and Merge`. Squashing removes detailed commit descriptions, so that `master` is not messy with every commit ever. Once the code has been squashed and merged, it is in the codebase and you have succeeded in contributing code to NUbots.
```
Comment (ysims) [inline src/book/04-general/01-git.mdx:43]:
```suggestion
Once the reviewers have approved the PR, and all of the comments are resolved, you can `Squash and Merge`. Squashing removes detailed commit descriptions, so that `master` is not messy with every commit ever. Once the code has been squashed and merged, it is in the codebase and you have succeeded in contributing code to NUbots.
```

Comment (ysims) [inline src/book/04-general/01-git.mdx:43]: Maybe mention formatting and no errors with buildkite, netlify, or whatever

File: src/book/04-general/01-git.mdx [hunk 13]
```diff
@@ -0,0 +1,59 @@
...
+
+## Commit Etiquette
+
+Commits should make minor changes, and the changes in each commit should have a common theme. A general mantra for contributing code is `commit often, push once`, but this isn't strict at all. Committing often means that the code is built up incrementally. Pushing once means that you don't spam all the other contributors with notifications.
+
+## Formatting and buildkite
+
+Before making a pull request, it is important that your code adheres to the formatting rules. For branches in the NUbots repo, running `./b format` before your last commit should fix formatting issues. Code in the NUbook repo must be formatted with `yarn format` before it can be committed.
```
Comment (ysims) [inline src/book/04-general/01-git.mdx:51]: Is there a similar thing for NUsight2 @JosephusPaye?

File: src/book/04-general/01-git.mdx [hunk 14]
```diff
@@ -0,0 +1,59 @@
...
+
+If your code is in a different repository and it uses C++ or protobuf files, use the `.clang-format` file in the NUbots repo to format them.
+
+If you get a message that your build failed, try running the relevant format commands and then committing the changes which they made. You won't be able to merge the PR until the build checks pass.
```
Comment (ysims) [inline src/book/04-general/01-git.mdx:55]:
```suggestion
If you get a message that your build failed, try running the relevant format commands and then committing any format changes. You won't be able to merge the PR until the build checks pass.
```

File: src/book/04-general/01-git.mdx [hunk 15]
```diff
@@ -0,0 +1,59 @@
...
+
+The reviewers will give constructive feedback about your code, by making comments on code snippets from your changes. You can respond to these on Github. They can also directly suggest changes to specific sections. Additional commits you make to the branch will update the PR and will reset their review of changed sections.
+
+Once the reviewers are happy with it, and all of the comments are resolved, you can `Squash and Merge`. Squashing removes detailed commit descriptions, so that `master` is not messy with every commit ever. Once the code has been squashed and merged, it is in the codebase and you have succeeded in contributing code to NUbots.
+
+## Commit Etiquette
+
+Commits should make minor changes, and the changes in each commit should have a common theme. A general mantra for contributing code is `commit often, push once`, but this isn't strict at all. Committing often means that the code is built up incrementally. Pushing once means that you don't spam all the other contributors with notifications.
+
+## Formatting and buildkite
```
Comment (ysims) [inline src/book/04-general/01-git.mdx:49]:
```suggestion
## Formatting and Buildkite
```

File: src/book/04-general/01-git.mdx (hunk 14 above)
Comment (ysims) [reply src/book/04-general/01-git.mdx:55]: Not sure on this one

File: src/book/04-general/01-git.mdx [hunk 16]
```diff
@@ -0,0 +1,59 @@
...
+
+Before making a pull request, it is important that your code adheres to the formatting rules. For branches in the NUbots repo, running `./b format` before your last commit should fix formatting issues. Code in the NUbook repo must be formatted with `yarn format` before it can be committed.
+
+If your code is in a different repository and it uses C++ or protobuf files, use the `.clang-format` file in the NUbots repo to format them.
+
+If you get a message that your build failed, try running the relevant format commands and then committing the changes which they made. You won't be able to merge the PR until the build checks pass.
+
+## Work in Progress PRs
+
+If you have code which isn't done yet, but which you would like feedback on, you can make a draft PR to get your code reviewed. Prepend `[WIP]` to the PR title when making draft pull requests. Draft PRs cannot be merged, so there is no need to worry about merging in bad code. Github has documentation about draft PRs [here.](#https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/about-pull-requests#draft-pull-requests)
```
Comment (ysims) [inline src/book/04-general/01-git.mdx:59]: there is a button you can press for this (drop down on 'create pull request' then change to draft pull request)

Comment (ysims) [inline src/book/04-general/01-git.mdx:59]:
```suggestion
If you have an incomplete project you would like feedback on, you can make a draft PR. Prepend `[WIP]` to the PR title when making draft pull requests. Draft PRs cannot be merged, so there is no need to worry about merging in bad code. GitHub has documentation about draft PRs [here.](#https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/about-pull-requests#draft-pull-requests)
```

File: src/book/04-general/01-git.mdx (hunk 13 above)
Comment (JosephusPaye) [reply src/book/04-general/01-git.mdx:51]: For NUsight, it's `yarn eslint` to find problems and `yarn eslint:fix` to fix.

File: src/book/04-general/01-git.mdx (hunk 16 above)
Comment (KipHamiltons) [reply src/book/04-general/01-git.mdx:59]: I was steering clear of explicitly saying which button and where the button is, but we can add if you reckon

File: src/book/04-general/01-git.mdx (hunk 6 above)
Comment (KipHamiltons) [reply src/book/04-general/01-git.mdx:11]: I have added which git commands are prerequisite, but I am not sure what to say about shell. It's mostly that you just can't be shellphobic lol

File: src/book/04-general/01-git.mdx (hunk 7 above)
Comment (KipHamiltons) [reply src/book/04-general/01-git.mdx:15]: I have made changes along these lines. let me know what u think.

File: src/book/04-general/01-git.mdx [hunk 17]
```diff
@@ -0,0 +1,61 @@
...
+slug: /guides/general/git
+---
+
+## Prerequisites
+
+This guide assumes you know a few basic shell commands, and the basics of git, such as adding, committing, and pushing to a repository.
+
+## Getting Code Into a Repository
+
+Let's suppose you have created a file, you have made which you would like to add to a NUbots repository, called `my_file.cpp`, which is part of your project to make the robot stand up.
```
Comment (Bidski) [inline src/book/04-general/01-git.mdx:15]:
```suggestion
Let's suppose you have created a file which you would like to add to a NUbots repository called `MakeRobotStand.cpp`, which is part of your project to make the robot stand up.
```

**Comment by JosephusPaye** [comment]
> It would be good for you to make a simple example and follow it through. [...]

Better yet, record a video? 😛😆

**Comment by Bidski** [comment]
I mean .... if you really want to lol

File: src/book/04-general/01-git.mdx (hunk 16 above)
Comment (ysims) [reply src/book/04-general/01-git.mdx]: Ye fair enough and you link to the GitHub page about it anyway

**Comment by KipHamiltons** [comment]
I do have a webcam... I could make a video if you think it would be super lit @JosephusPaye

**Comment by JosephusPaye** [comment]
I do 😄

File: src/book/03-guides/04-general/01-git.mdx [hunk 18]
```diff
@@ -0,0 +1,61 @@
...
+
+Where `<branch name>` corresponds to the branch you found in the last step.
+
+If you are not working on an existing project, you need to make a branch to commit your changes to. The format for branch names is `surname/project_name`. So for this example it could be `citizen/stand_up`.
+
+### Making a Pull Request
+
+Once you have a set of commits on your branch which you think should be added to the main codebase, you can make a Pull Request (PR). PRs are where users ask the repository to pull their branch into the `master` branch.
+
+Github has documentation about PRs [here.](#https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/creating-a-pull-request)
```
Comment (ysims) [inline src/book/03-guides/04-general/01-git.mdx:35]:
```suggestion
GitHub has documentation about PRs [here.](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/creating-a-pull-request)
```

File: src/book/03-guides/04-general/01-git.mdx [hunk 19]
```diff
@@ -0,0 +1,61 @@
...
+
+To be approved, you should add reviewers to the pull request. If you have a project mentor, they should be a reviewer, as well as someone else who knows about that part of the codebase. Once the reviewers have approved the changes, the changes can be merged into the master branch.
+
+### Code Review
+
+The reviewers will give constructive feedback about your code, by making comments on code snippets from your changes. You can respond to these on GitHub. They can also directly suggest changes to specific sections. Additional commits you make to the branch will update the PR and will reset reviews of the changed sections.
+
+Once all of the comments are resolved, the reviewers have approved the PR, and the build checks have passed (see [_Formatting and Buildkite Checks_](#formatting-and-buildkite-checks)), you can `Squash and Merge`. Squashing removes detailed commit descriptions, so that `master` is not messy with every commit ever. Once the code has been squashed and merged, it is in the codebase and you have succeeded in contributing code to NUbots.
```
Comment (ysims) [inline src/book/03-guides/04-general/01-git.mdx:45]:
```suggestion
Once all of the comments are resolved, the reviewers have approved the PR, and the build checks have passed (see [_Formatting and Buildkite Checks_](#formatting-and-buildkite-checks)), you can `Squash and Merge`. Squashing removes detailed commit descriptions so that the `master` branch history is clean and easy to follow. Once the code has been squashed and merged, it is in the codebase and you have succeeded in contributing code to NUbots.
```

File: src/book/03-guides/04-general/01-git.mdx [hunk 20]
```diff
@@ -0,0 +1,61 @@
...
+
+To be approved, you should add reviewers to the pull request. If you have a project mentor, they should be a reviewer, as well as someone else who knows about that part of the codebase. Once the reviewers have approved the changes, the changes can be merged into the master branch.
+
+### Code Review
+
+The reviewers will give constructive feedback about your code, by making comments on code snippets from your changes. You can respond to these on GitHub. They can also directly suggest changes to specific sections. Additional commits you make to the branch will update the PR and will reset reviews of the changed sections.
+
+Once all of the comments are resolved, the reviewers have approved the PR, and the build checks have passed (see [_Formatting and Buildkite Checks_](#formatting-and-buildkite-checks)), you can `Squash and Merge`. Squashing removes detailed commit descriptions so that the `master` branch history is clean and easy to follow. Once the code has been squashed and merged, it is in the codebase and you have succeeded in contributing code to NUbots.
```
Comment (JosephusPaye) [inline src/book/03-guides/04-general/01-git.mdx:45]: Not code 😛 + wording

```suggestion
Once all of the comments are resolved, the reviewers have approved the PR, and the build checks have passed (see [_Formatting and Buildkite Checks_](#formatting-and-buildkite-checks)), you can **Squash and Merge**. Squashing combines every commit on the PR into a single commit so that the `master` branch history is clean and easy to follow. Once the code has been squashed and merged, it is in the codebase and you have succeeded in contributing code to NUbots.
```

File: src/book/03-guides/04-general/01-git.mdx [hunk 21]
```diff
@@ -0,0 +1,61 @@
...
+
+Before making a pull request, it is important that your code adheres to the formatting rules. For branches in the NUbots repo, running `./b format` before your last commit should fix formatting issues. Code in the NUbook repo must be formatted with `yarn format` before it can be committed. For NUsight code, use `yarn eslint` to find problems, and `yarn eslint:fix` to fix them.
+
+If your code is in a different repository and it uses C++ or protobuf files, use the `.clang-format` file in the NUbots repo to format them.
+
+If you get a message that your build failed, try running the relevant format commands and then committing any format changes. You won't be able to merge the PR until the build checks pass.
+
+## Work in Progress PRs
+
+If you have an incomplete project you would like feedback on, you can make a draft PR to get your code reviewed. Prepend `[WIP]` to the PR title when making draft pull requests. Draft PRs cannot be merged, so there is no need to worry about merging in bad code. GitHub has documentation about draft PRs [here.](#https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/about-pull-requests#draft-pull-requests)
```
Comment (JosephusPaye) [inline src/book/03-guides/04-general/01-git.mdx:61]: Make links describe what they're pointing to:

```suggestion
If you have an incomplete project you would like feedback on, you can make a draft PR to get your code reviewed. Prepend `[WIP]` to the PR title when making draft pull requests. Draft PRs cannot be merged, so there is no need to worry about merging in bad code. GitHub has [documentation about draft PRs here](#https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/about-pull-requests#draft-pull-requests).
```

File: src/book/03-guides/04-general/01-git.mdx [hunk 22]
```diff
@@ -0,0 +1,61 @@
...
+
+Where `<branch name>` corresponds to the branch you found in the last step.
+
+If you are not working on an existing project, you need to make a branch to commit your changes to. The format for branch names is `surname/project_name`. So for this example it could be `citizen/stand_up`.
+
+### Making a Pull Request
+
+Once you have a set of commits on your branch which you think should be added to the main codebase, you can make a Pull Request (PR). PRs are where users ask the repository to pull their branch into the `master` branch.
+
+GitHub has documentation about PRs [here.](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/creating-a-pull-request)
```
Comment (JosephusPaye) [inline src/book/03-guides/04-general/01-git.mdx:35]: Make links describe what they're pointing to:

```suggestion
GitHub has [documentation about PRs here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/creating-a-pull-request).
```

File: src/book/03-guides/04-general/01-git.mdx [hunk 23]
```diff
@@ -0,0 +1,61 @@
...
+
+If you are working within an existing project, then you need to find out which branch the work is on. To list the branches, use the command `git branch -r`. When you have found the branch you want to push to, swap to it using the following command:
+
+```sh
+git checkout <branch name>
+```
+
+Where `<branch name>` corresponds to the branch you found in the last step.
+
+If you are not working on an existing project, you need to make a branch to commit your changes to. The format for branch names is `surname/project_name`. So for this example it could be `citizen/stand_up`.
```
Comment (JosephusPaye) [inline src/book/03-guides/04-general/01-git.mdx:29]:
```suggestion
If you are not working on an existing project, you need to make a branch to commit your changes to. The format for branch names is `your_surname/brief_description_of_project`. So if your surname is Hamiltons, and your project is making the robot stand up, a good branch name would be `hamiltons/stand_up`.
```

File: src/book/03-guides/04-general/01-git.mdx [hunk 24]
```diff
@@ -0,0 +1,61 @@
...
+
+### Code Review
+
+The reviewers will give constructive feedback about your code, by making comments on code snippets from your changes. You can respond to these on GitHub. They can also directly suggest changes to specific sections. Additional commits you make to the branch will update the PR and will reset reviews of the changed sections.
+
+Once all of the comments are resolved, the reviewers have approved the PR, and the build checks have passed (see [_Formatting and Buildkite Checks_](#formatting-and-buildkite-checks)), you can `Squash and Merge`. Squashing removes detailed commit descriptions so that the `master` branch history is clean and easy to follow. Once the code has been squashed and merged, it is in the codebase and you have succeeded in contributing code to NUbots.
+
+## Commit Etiquette
+
+Commits should make minor changes, and the changes in each commit should have a common theme. A general mantra for contributing code is `commit often, push once`, but this isn't strict at all. Committing often means that the code is built up incrementally. Pushing once means that you don't spam all the other contributors with notifications.
```
[the remark this answers is not in the document]
Comment (KipHamiltons) [reply src/book/03-guides/04-general/01-git.mdx:51]: hahaha fair enough, I just like that style of emphasis :p

File: src/book/03-guides/04-general/01-git.mdx (hunk 23 above)
Comment (KipHamiltons) [reply src/book/03-guides/04-general/01-git.mdx:31]: Are we usually using real people's names in examples in NUbook?

Comment (JosephusPaye) [reply src/book/03-guides/04-general/01-git.mdx:31]: Yep :P

Comment (KipHamiltons) [reply src/book/03-guides/04-general/01-git.mdx:31]: I dunno, it just seemed kinda like, who on Earth would be named Hamiltons anyway??? stupid name :laughing:
